src/datasources/email/email.datasource.ts

```diff
@@ -0,0 +1,104 @@
+import { Inject, Injectable } from '@nestjs/common';
+import * as postgres from 'postgres';
+import { IEmailDataSource } from '@/domain/interfaces/email.datasource.interface';
+import { EmailVerificationCode } from '@/domain/email/entities/email-verification-code.entity';
+import { EmailAddressDoesNotExistError } from '@/datasources/email/errors/email-address-does-not-exist.error';
+import { Email } from '@/datasources/email/entities/email.entity';
+
+@Injectable()
+export class EmailDataSource implements IEmailDataSource {
+  constructor(@Inject('DB_INSTANCE') private readonly sql: postgres.Sql) {}
+
+  private async _getEmail(args: {
+    chainId: string;
+    safeAddress: string;
+    signer: string;
+  }): Promise<Email> {
+    const [email] = await this.sql<Email[]>`SELECT *
+                                            FROM emails.signer_emails
+                                            WHERE chain_id = ${args.chainId}
+                                              and safe_address = ${args.safeAddress}
+                                              and signer = ${args.signer}`;
+
+    if (!email) {
+      throw new EmailAddressDoesNotExistError(
+        args.chainId,
+        args.safeAddress,
+        args.signer,
+      );
+    }
+
+    return email;
+  }
+
+  async saveEmail(args: {
+    chainId: string;
+    safeAddress: string;
+    emailAddress: string;
+    signer: string;
+    code: number;
+  }): Promise<EmailVerificationCode> {
+    return await this.sql.begin(async (sql) => {
+      const [email] = await sql<Email[]>`
+          INSERT INTO emails.signer_emails (chain_id, email_address, safe_address, signer, verification_code)
+          VALUES (${args.chainId}, ${args.emailAddress}, ${args.safeAddress}, ${args.signer}, ${args.code})
+          RETURNING *
+      `;
+
+      return <EmailVerificationCode>{
+        emailAddress: email.email_address,
+        verificationCode: email.verification_code,
+      };
+    });
+  }
+
+  async setVerificationCode(args: {
+    chainId: string;
+    safeAddress: string;
+    signer: string;
+    code: number;
+  }): Promise<EmailVerificationCode> {
+    const [email] = await this.sql<Email[]>`UPDATE emails.signer_emails
+                                            SET verification_code    = ${args.code},
+                                                verification_sent_on = now()
```

Comment:
Nitpick: There is no need to tackle this on the PR, but we could consider setting this value (`verification_sent_on`) once the message is actually sent to the user's email. That would also provide insights to detect errors (i.e.: a null value here would indicate the email couldn't be sent for some reason). What do you think?

Reply:
I've decided to tackle it here. I agree that decoupling the value that is set here and allowing it to be `NULL` would represent better the state of the verifications status 🙌  [641cb20](https://github.com/safe-global/safe-client-gateway/pull/829/commits/641cb204025e4ca4b2373dd5904ad99710b7da68)